### ISSUE TITLE
Hook s3-credentials-relation-changed failed

### ISSUE THREAD

**Issue by rgildein** [canonical/spark-history-server-k8s-operator#56]
<!-- Thank you for submitting a bug report! All fields are required unless marked optional. -->

## Steps to reproduce
Run spark bundle test from [spark-k8s-bundle](https://github.com/canonical/spark-k8s-bundle) repo.

1. juju --channel 3.4/stable
1. microk8s --channel  1.28-strict/stable
1. IPADDR=$(ip -4 -j route get 2.2.2.2 | jq -r '.[] | .prefsrc') 
1. microk8s enable rbac dns minio metallb:$IPADDR-$IPADDR
1. cd python && tox run -r -e integration-sparkjob -- -m 'not unstable' --backend yaml --cos-model cos --spark-version "3.4.2" --keep-models --debug

## Expected behavior

All applications in `test-spark-job-...` model in active idle.

## Actual behavior
```bash
Model                Controller          Cloud/Region        Version  SLA          Timestamp
test-spark-job-9ywv  microk8s-localhost  microk8s/localhost  3.4.6    unsupported  10:11:14Z

SAAS                             Status  Store               URL
grafana-dashboards               active  microk8s-localhost  admin/cos.grafana-dashboards
loki-logging                     active  microk8s-localhost  admin/cos.loki-logging
prometheus-receive-remote-write  active  microk8s-localhost  admin/cos.prometheus-receive-remote-write
traefik                          active  microk8s-localhost  admin/cos.traefik

App                Version  Status   Scale  Charm                         Channel        Rev  Address         Exposed  Message
agent              0.35.2   active       1  grafana-agent-k8s             latest/stable   45  10.152.183.165  no
cos-configuration  3.5.0    active       1  cos-configuration-k8s         latest/edge     61  10.152.183.210  no
history-server              waiting      1  spark-history-server-k8s      3.4/edge        30  10.152.183.230  no       installing agent
integration-hub             active       1  spark-integration-hub-k8s     latest/edge     22  10.152.183.154  no
kyuubi                      active       3  kyuubi-k8s                    latest/edge     27  10.152.183.76   no
kyuubi-users       14.11    active       1  postgresql-k8s                14/stable      281  10.152.183.219  no
metastore          14.11    active       1  postgresql-k8s                14/stable      281  10.152.183.188  no
pushgateway        1.7.0    active       1  prometheus-pushgateway-k8s    latest/edge     16  10.152.183.35   no
s3                          active       1  s3-integrator                 latest/edge     17  10.152.183.254  no
scrape-config      n/a      active       1  prometheus-scrape-config-k8s  latest/stable   51  10.152.183.78   no
zookeeper          3.8.4    active       3  zookeeper-k8s                 3/edge          59  10.152.183.231  no

Unit                  Workload  Agent  Address       Ports  Message    
agent/0*              active    idle   10.1.202.221      
cos-configuration/0*  active    idle   10.1.202.244      
history-server/0*     error     idle   10.1.202.217         hook failed: "s3-credentials-relation-changed" 
integration-hub/0*    active    idle   10.1.202.214      
kyuubi-users/0*       active    idle   10.1.202.225         Primary    
kyuubi/0              active    idle   10.1.202.218      
kyuubi/1*             active    idle   10.1.202.201      
kyuubi/2              active    idle   10.1.202.219      
metastore/0*          active    idle   10.1.202.247         Primary    
pushgateway/0*        active    idle   10.1.202.243      
s3/0*                 active    idle   10.1.202.207      
scrape-config/0*      active    idle   10.1.202.246  
```

## Versions

<!-- Run `lsb_release -sd` -->
Operating system:  ubuntu-22.04

<!-- Run `juju version` -->
Juju CLI:  juju --version 3.4.6-genericlinux-amd64

<!-- Model version from `juju status` -->
Juju agent: 3.4.6 

<!-- App revision from `juju status` or (advanced) commit hash -->
Charm revision:  30

<!-- Run `microk8s version` -->
microk8s:  1.28

## Log output
<!-- Please enable debug logging by running `juju model-config logging-config="<root>=INFO;unit=DEBUG"` (if possible) -->
<!-- Then, run `juju debug-log --replay > log.txt` and upload "log.txt" file here -->
Juju debug log: 

```bash
unit-history-server-0: 10:12:11 INFO unit.history-server/0.juju-log s3-credentials:38: S3 Credentials changed                                                                                                                                                                             
unit-history-server-0: 10:12:11 INFO unit.history-server/0.juju-log s3-credentials:38: Layer: {"summary": "spark history server layer", "description": "pebble config layer for spark history server", "services": {"history-server": {"environment": {"SPARK_DAEMON_JAVA_OPTS": "-javaage
nt:/opt/spark/jars/jmx_prometheus_javaagent-0.20.0.jar=9101:/etc/spark/conf/jmx_prometheus.yaml", "SPARK_PROPERTIES_FILE": "/etc/spark/conf/spark-properties.conf", "SPARK_HISTORY_OPTS": ""}}}}                                                                                          
unit-history-server-0: 10:12:12 ERROR unit.history-server/0.juju-log s3-credentials:38: Uncaught exception while in charm code:                                                                                                                                                           
Traceback (most recent call last):                                                                                                                                                                                                                                                        
  File "/var/lib/juju/agents/unit-history-server-0/charm/./src/charm.py", line 57, in <module>                                                                                                                                                                                            
    main(SparkHistoryServerCharm)                                                                                                                                                                                                                                                         
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/main.py", line 553, in main                                                                                                                                                                                             
    manager.run()                                                                                                                                                                                                                                                                         
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/main.py", line 529, in run
    self._emit()                                                                                                                                                                                                                                                                  [0/4556]
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/main.py", line 518, in _emit                                                                                                                                                                                            
    _emit_charm_event(self.charm, self.dispatcher.event_name, self._juju_context)                                                                                                                                                                                                         
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/main.py", line 139, in _emit_charm_event                                   
    event_to_emit.emit(*args, **kwargs)                                                                                                                                                                                                                                                   
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/framework.py", line 347, in emit                                                                                                                                                                                        
    framework._emit(event)                                                                                                                                                                                                                                                                
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/framework.py", line 853, in _emit                                          
    self._reemit(event_path)                                                                                                                                                                                                                                                              
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/framework.py", line 943, in _reemit                                        
    custom_handler(event)                                                                                                                    
  File "/var/lib/juju/agents/unit-history-server-0/charm/lib/charms/data_platform_libs/v0/s3.py", line 768, in _on_relation_changed          
    getattr(self.on, "credentials_changed").emit(                                                                                                                                                                                                                                         
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/framework.py", line 347, in emit                                                                                                                                                                                        
    framework._emit(event)                                                                                                                   
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/framework.py", line 853, in _emit                                                                                                                                                                                       
    self._reemit(event_path)                                                                                                                                                                                                                                                              
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/framework.py", line 943, in _reemit                                                                                                                                                                                     
    custom_handler(event)                                                                                                                    
  File "/var/lib/juju/agents/unit-history-server-0/charm/src/events/base.py", line 66, in wrapper_hook                                                                                                                                                                                    
    res = hook(event_handler, event)                                                                                                         
  File "/var/lib/juju/agents/unit-history-server-0/charm/src/events/base.py", line 96, in wrapper_hook                                       
    return hook(event_handler, event)                                                                                                        
  File "/var/lib/juju/agents/unit-history-server-0/charm/src/events/s3.py", line 44, in _on_s3_credential_changed                                                                                                                                                                         
    self.history_server.update(                                                                                                                                                                                                                                                           
  File "/var/lib/juju/agents/unit-history-server-0/charm/src/managers/history_server.py", line 183, in update                                
    self.workload.start()                                                                                                                                                                                                                                                                 
  File "/var/lib/juju/agents/unit-history-server-0/charm/src/workload.py", line 89, in start                                                                                                                                                                                              
    self.container.restart(self.HISTORY_SERVER_SERVICE)                                                                                                                                                                                                                                   
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/model.py", line 2293, in restart                                           
    self._pebble.restart_services(service_names)                                                                                                                                                                                                                                          
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/pebble.py", line 2203, in restart_services                                 
    return self._services_action('restart', services, timeout, delay)                                                                        
  File "/var/lib/juju/agents/unit-history-server-0/charm/venv/ops/pebble.py", line 2228, in _services_action                                 
    raise ChangeError(change.err, change)                                                                                                                                                                                                                                                 
ops.pebble.ChangeError: cannot perform the following tasks:                                                                                                                                                                                                                               
- Start service "history-server" (cannot start service: exited quickly with code 134)                                                        
----- Logs from task 0 -----                                                                                                                 
2024-11-14T10:12:11Z INFO Service "history-server" has never been started.                                                                                                                                                                                                                
----- Logs from task 1 -----                                                                                                                                                                                                                                                              
2024-11-14T10:12:11Z INFO Most recent service output:                                                                                                                                                                                                                                     
    (...)                                                                                                                                                                                                                                                                                 
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                          
        at java.base/java.lang.reflect.Method.invoke(Method.java:569)                                                                        
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:491)                           
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:503)                          
    Caused by: java.lang.NoSuchMethodError: 'io.prometheus.jmx.shaded.io.prometheus.client.Collector io.prometheus.jmx.BuildInfoCollector.register()'                                                                                                                                     
        at io.prometheus.jmx.JavaAgent.premain(JavaAgent.java:54)                                                                                                                                                                                                                         
        ... 6 more                                                                                                                           
    *** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message Outstanding error when calling method in invokeJavaAgentMainMethod at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 619                                                         
    *** java.lang.instrument ASSERTION FAILED ***: "success" with message invokeJavaAgentMainMethod failed at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 459                                                                                                     
    *** java.lang.instrument ASSERTION FAILED ***: "result" with message agent load/premain call failed at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 422                                                                                                        
    FATAL ERROR in native method: processing of -javaagent failed, processJavaStart failed                                                   
    Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)                                                                                                                                                                                                       
    V  [libjvm.so+0x8c642e]  jni_FatalError+0xbe                                                                                             
    V  [libjvm.so+0xa4492f]  JvmtiExport::post_vm_initialized()+0xcf                                                                         
    V  [libjvm.so+0xeb9e41]  Threads::create_vm(JavaVMInitArgs*, bool*)+0x921                                                                                                                                                                                                             
    V  [libjvm.so+0x8e02c5]  JNI_CreateJavaVM+0x55                                                                                                                                                                                                                                        
    C  [libjli.so+0x3d43]  JavaMain+0x93                                                                                                                                                                                                                                                  
    C  [libjli.so+0x7fbd]  ThreadJavaMain+0xd                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                          
    /opt/spark/sbin/spark-daemon.sh: line 133:  1966 Aborted                 (core dumped) "$@"                                                                                                                                                                                           
2024-11-14T10:12:11Z ERROR cannot start service: exited quickly with code 134                                                                
-----                                                                                                                                        
unit-history-server-0: 10:12:12 ERROR juju.worker.uniter.operation hook "s3-credentials-relation-changed" (via hook dispatching 
```

<!-- (Optional) Copy the logs that are relevant to the bug & paste inside triple backticks below -->


## Additional context
<!-- (Optional) Add any additional information here -->

**Comment by rgildein** [comment]
related with: https://github.com/canonical/spark-k8s-bundle/issues/53

**Comment by welpaolo** [comment]
Solved by #57
